external-crates/move/move-symbol-pool/src/symbol.rs

```diff
@@ -142,17 +249,128 @@ mod tests {
+            let sym = Symbol::from(&s[..n]);
+            assert_eq!(sym.len(), n);
+            let expected_tag = if n <= MAX_INLINE_LEN {
+                Tag::Inlined
+            } else {
+                Tag::Dynamic
+            };
+            assert_eq!(sym.tag(), expected_tag)
+        }
+    }
+
     #[test]
     fn test_from_different_strings_have_different_addresses() {
         let s1 = Symbol::from("hi");
         let s2 = Symbol::from("hello");
+        assert_eq!(s1.tag(), Tag::Inlined);
+        assert_eq!(s2.tag(), Tag::Inlined);
+        assert_ne!(s1.0, s2.0);
+
+        let s1 = Symbol::from("hi");
+        let s2 = Symbol::from("bonjour le monde");
+        assert_eq!(s1.tag(), Tag::Inlined);
+        assert_eq!(s2.tag(), Tag::Dynamic);
+        assert_ne!(s1.0, s2.0);
+
+        let s1 = Symbol::from("bonjour!");
+        let s2 = Symbol::from("bonjour le monde");
+        assert_eq!(s1.tag(), Tag::Dynamic);
+        assert_eq!(s2.tag(), Tag::Dynamic);
         assert_ne!(s1.0, s2.0);
     }
 
     #[test]
     fn test_from_identical_strings_have_the_same_address() {
-        let s1 = Symbol::from("bonjour");
-        let s2 = Symbol::from("bonjour");
+        let s1 = Symbol::from("hi");
+        let s2 = Symbol::from("hi");
+        assert_eq!(s1.tag(), Tag::Inlined);
+        assert_eq!(s2.tag(), Tag::Inlined);
+        assert_eq!(s1.0, s2.0);
+
+        let s1 = Symbol::from("bonjour le monde");
+        let s2 = Symbol::from("bonjour le monde");
+        assert_eq!(s1.tag(), Tag::Dynamic);
+        assert_eq!(s2.tag(), Tag::Dynamic);
         assert_eq!(s1.0, s2.0);
     }
+
+    #[test]
+    fn test_all_static_strings() {
+        // check all symbols round trip correctly
+        for s in STATIC_SYMBOLS {
+            let from_s = Symbol::from(*s);
+            let idx = *STATIC_SYMBOL_IDX.get(*s).unwrap();
+            let from_idx = Symbol::pack_static(idx);
+            assert_eq!(from_s.tag(), Tag::Static);
+            assert_eq!(from_idx.tag(), Tag::Static);
+            assert_eq!(from_s.0, from_idx.0);
+            assert_eq!(from_s.0, from_idx.0);
+            assert_eq!(from_s, from_idx);
+        }
+        // check all entries in `STATIC_SYMBOL_IDX` are present in `STATIC_SYMBOLS`
+        let mut entries = STATIC_SYMBOL_IDX.entries().collect::<Vec<_>>();
+        entries.sort_by_key(|(_, idx)| *idx);
+        assert_eq!(entries.len(), STATIC_SYMBOLS.len());
+        for ((map_s, map_idx), (sym_idx, sym_s)) in
+            entries.iter().zip(STATIC_SYMBOLS.iter().enumerate())
+        {
+            let map_s = *map_s;
+            let map_idx = **map_idx as usize;
+            assert_eq!(map_idx, sym_idx);
+            assert_eq!(map_s, sym_s);
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! static_symbols_with_idx {
+    ($($str:tt => $idx:expr),* $(,)?) => {
+        #[macro_export]
+        macro_rules! symbol {
+            $(($str) => { $crate::symbol::Symbol::pack_static($idx) };)*
```

Comment:
IIUC, this means that even though we call it like this: `symbol!("foo")`, if `"foo"` isn't one of the pre-allocated static symbols, the macro will fail to expand, is that right? If so, what's the benefit of the `symbol!` macro over just assigning a bunch of associated consts?:

```rust
impl Symbol {
    const STD: Symbol = ...;
    const VECTOR: Symbol = ...;
    // ...
}
```

Reply:
P.S. you know things are getting serious when your macro expansion contains another macro definition...

Reply:
I think just for readability for me personally. I think `n == symbol!("std")` just reads a bit nicer than `n == Symbol::STD`, and more closely mirrors `n.as_str() == "std"`

(just as a note since my examples are silly, this is more necessary when you need to start adding/removing/borrowing map entries)

Reply:
And fwiw it keeps us close to `string_cache`, which our symbol pool is basically a fork, that just leaks instead 

https://doc.servo.org/string_cache/index.html

They have this notion of static "atoms", which they use the syntax `atom!("foo")` for

Reply:
Though I think there is a tradeoff here, we could make a default case of `($s:expr) => Symbol::from`? 
You will still get an error, but only in `const` scenarios if isn't static

Reply:
The distinction for me between `symbol!` (as it is now) and `atom!` for me is precisely that `atom!` can be called anywhere because it leaks the storage for the static element, whereas `symbol!("x")`  behaves identically to accessing a binding for `x` (it will complain if you haven't declared it), but with what I expect would be quite a roundabout error -- thus the question about just using a constant.

If we can get `symbol!` to have a fallback case that is desirable, then I'm all for keeping this syntax, but it seems to me like the behaviour you're trying to encourage is for people to declare more static symbols if they are missing?  In which case, I would advocate for having the macro define constants -- it makes it clearer that they don't appear out of thin air because of the call.

Reply:
I'll add the fallback case then

Reply:
In the case where a static hasn't been defined, is it preferable to fall back to `Symbol::from` or to add the symbol to the set of predefined statics? I could see it going either way, given that you'll get a complaint in `const` contexts.  If the latter though, it seems worthwhile to make the macro define constants rather than another macro.

Reply:
It will call `Symbol::from`, it will be unable to add it to the predefined statics. If I could get that to work, I wouldn't need the list of predefined statics (which would be awesome!)

Reply:
I understand that bit, I meant separately from what could be automated, what's the best fallback action; (automatically) fall back to the dynamic `Symbol::from` or produce an error that would prompt someone to (manually) add the symbol to the predefined statics, even in non-const contexts?

Reply:
I misunderstood your previous comment then, sorry. 

The current state of the PR does the "automatic" solution with the fall back to `Symbol::from`. But I would prefer the "manual" solution with the prompt, since it feels strange to have a static symbol that isn't in the static pool.
Would something like this work for you?
```
error[[E0080]](https://doc.rust-lang.org/stable/error_codes/E0080.html): evaluation of constant value failed
  --> src/lib.rs:10:4
   |
10 |    symbol!("foo");
   |    ^^^^^^^^^^^^^^ the evaluated program panicked at 'Unbound static symbol: "foo". Add it to the `static_symbols` list in `<path to file assuming I get the macro to work>`. Or use `Symbol::from`', src/lib.rs:10:4
```

Reply:
That seems better, but it feels as though we're reinventing variable bindings, using strings and macros in a way where we get non-standard error message and don't get auto-complete support from the language server, whereas if the macro defined a bunch of constants under an impl block for `Symbol`, you would get the same behaviour but in a more standard way -- that's what I was trying to get across in my initial comment.

I think we understand each other now though, and ultimately, I don't think I'll spend as much time in this codebase as you, so if you still prefer working with `symbol!("...")` over `Symbol::...` then I won't block this PR for it!

Reply:
To me, the constants will be rather annoying, mostly just from a readability standpoint. Imagine if you had `Symbol::SPEC_DOMAIN`, and then you go and look up at the definition of the constant, one of a few things will happen:
- Your editor will not be able to look at the definition (from the macros)
- You will just see `Symbol::pack_static((0 + 1 + 1 + ... + 1) - (0 + 1 + ... + 1))`

I'm not really sure that's readable. 

I would argue we aren't reinventing names, we are reinventing `str`. The whole `Symbol` system is basically a custom built string library, and we want literals in that library for the same reasons you would want literals normally. And as such, I don't think it would really be expected to have autocomplete or anything, you just write out your literal and thats that

Reply:
I'm pretty sure the editor would have taken you to the macro call that generated the definition similar to the custom error message you suggested but automatic, so by suggesting with the macro syntax that it's defining bindings (e.g. with syntax like `static_symbols! { STD = "std"; FOO = "foo"; ... }`) , you could make it clear that they need to add another.

On reinventing strings rather than bindings, I go back to the argument about `atom!` vs `symbol!` -- string literals (and `atom!`) don't require declarations, whereas bindings (and `symbol!`) do, and at point of use it feels more important to me to make it clear that you need to declare the symbol to use it, than to reinforce the similarity with strings (again, story would be different if the `symbol!` macro introduced the symbol as well as used it).

Reply:
That is not true for `atom`, it requires being added to the list in the `build.rs`

Reply:
😮, okay well in that case ... I'm also perturbed by `atom!`'s design.  If it requires a `build.rs`, I would have thought they could have just scraped the source code for calls to gather which strings were necessary?

Reply:
> I would have thought they could have just scraped the source code for calls to gather which strings were necessary?

That would be really nice. And if it did that, I would have forked `string_cache_codegen` instead of doing this zany macro.